### PR TITLE
Default to utf8mb4

### PR DESCRIPTION
**Pull request by Alkarex**
https://github.com/friendica/friendica/issues/3324

### CODE REVIEW

**Comment by annando** [comment]
In new installations there **should** be the line ```$a->config['system']['db_charset'] = "utf8mb4";``` in the .htconfig.php - so new installations **should** be okay (which doesn't seem so).

Old installations still are on "utf8" - with the connection mostly defaulting to "latin1" - so the question is if this fix would cause problems to these installations.

When this patch would be accepted, we could have a mix in charsets and collations in our database.

**Comment by Hypolite** [comment]
Do you mean older installs never get converted to `utf8mb4`?

**Comment by annando** [comment]
I'm unsure how to manage this the best way. Old installations could be in any state. I don't think that we could create some full automatic convert script.

**Comment by Alkarex** [comment]
An update to utf8mb4 (if supported) should work from whichever state: https://github.com/friendica/friendica/issues/3324#issuecomment-293599276

**Comment by annando** [comment]
We had to make sure that the dbstructure.php would do the changes in the collation automatically.

**Comment by Hypolite** [comment]
Are we doing it right now?

**Comment by Hypolite** [comment]
Should we reject this PR until we have a strong update process?

**Comment by Alkarex** [comment]
This only impacts new installations, isn't it? In which case would utf8 be a better choice than utf8mb4?

**Comment by annando** [comment]
You are the better SQL expert, I think. So the question is if there could be any problems with mixed tables.

**Comment by annando** [comment]
@Alkarex I want explain the problem that we are having on most old installations. See this post: https://pirati.ca/display/ec054ce71958efc4183486f210699509

In the database it looks like this:
```
> select body from item where guid="ec054ce71958efc4183486f210699509" and uid=0;
+---------------------------------------------------------------------------------------------------------------------------------------+
| body                                                                                                                                  |
+---------------------------------------------------------------------------------------------------------------------------------------+
| Wann begreifen die Menschen eigentlich endlich das Prinzip der Massenträgheit am Beispiel der Flaschen auf dem Supermarktfließband?   |
+---------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

> 
```

Now we look at another server: https://squeet.me/display/ec054ce71958efc4183486f210699509

As expected, it looks the same. But now have a look at the database:
```
> select body from item where guid="ec054ce71958efc4183486f210699509" and uid=0;
+-------------------------------------------------------------------------------------------------------------------------------------------+
| body                                                                                                                                      |
+-------------------------------------------------------------------------------------------------------------------------------------------+
| Wann begreifen die Menschen eigentlich endlich das Prinzip der MassentrÃ¤gheit am Beispiel der Flaschen auf dem SupermarktflieÃŸband?     |
+-------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

>
```
The first database behaves like newly installed ones, the second one behaves like the old ones. (Both are on the same system)

So when we do anything to convert data, we have to make sure that it doesn't destroys the content.

**Comment by annando** [comment]
I will do some tests and I guess I will rework the dbstructure.php in total.

**Comment by Alkarex** [comment]
As far as I can see:
* New installations should be full utf8mb4 (tables + communication)
* Old installations with full utf8 (tables + communication) can be safely updated to utf8mb4
* Other cases :-)

Could you try to get a `SHOW CREATE TABLE item;` from your second example (the problematic one)? I am guessing the data is stored in UTF-8 (one can see it is multi-byte) but the communication was not.

**Comment by annando** [comment]
In the second example the table is simply utf8, not utf8mb4. The communication between php and the database server is done in latin1, since this is the default.

**Comment by annando** [comment]
I did some tests and a new branch. I will now do a test on one of my machines. Later this evening (or this weekend) I will do a pull request for that. This will then make this PR obsolete.

My changes are looking like this: https://github.com/annando/friendica/commit/fdcb6d3c6f785ea588f3ce7db2659764032c114d

**Comment by Hypolite** [comment]
I reviewed your changes, and I noticed you didn't use `CONVERT` in your `ALTER TABLE` statements. Is the data automatically converted when you change a field charset/collation?

**Comment by annando** [comment]
The ```CONVERT``` would automatically change field types. "text" becomes "mediumtext" and so on. I'm doing it field by field. This can be done as well and is equivalent in the end result.

**Comment by Hypolite** [comment]
Is this rendered obsolete by #3341 ?

**Comment by Alkarex** [comment]
I have updated this PR to replace other instances of UT8 by UTF8MB4

File: mod/install.php [hunk 1]
```diff
@@ -48,7 +48,7 @@ function install_post(App $a) {
 				$db = new dba($dbhost, $dbuser, $dbpass, '', true);
 
 				if(! get_db_errno()) {
-					$r = q("CREATE DATABASE '%s'",
+					$r = q("CREATE DATABASE '%s' DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci",
```
Comment (annando) [inline mod/install.php]: It is enough to simply define the collation.

File: library/Smarty/demo/plugins/resource.mysql.php [hunk 2]
```diff
@@ -10,7 +10,7 @@
  *   `modified` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  *   `source` text,
  *   PRIMARY KEY (`name`)
- * ) ENGINE=InnoDB DEFAULT CHARSET=utf8;</pre>
+ * ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;</pre>
```
Comment (annando) [inline library/Smarty/demo/plugins/resource.mysql.php]: We shouldn't change stuff in third party libraries if not necessary.

File: library/Smarty/demo/plugins/resource.mysqls.php [hunk 3]
```diff
@@ -12,7 +12,7 @@
  *   `modified` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  *   `source` text,
  *   PRIMARY KEY (`name`)
- * ) ENGINE=InnoDB DEFAULT CHARSET=utf8;</pre>
+ * ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;</pre>
```
Comment (annando) [inline library/Smarty/demo/plugins/resource.mysqls.php]: Same goes here.

File: view/theme/frio/img/dynamic.php [hunk 4]
```diff
@@ -7,7 +7,7 @@
       'dsn' => 'mysql:host=127.0.0.1;dbname=installer',
       'username' => 'root',
       'password' => 'root',
-      'charset' => 'utf8',
+      'charset' => 'utf8mb4',
```
Comment (annando) [inline view/theme/frio/img/dynamic.php]: @rabuzarus where is this file needed for?

File: library/Smarty/demo/plugins/resource.mysql.php (hunk 2 above)
Comment (Alkarex) [inline library/Smarty/demo/plugins/resource.mysql.php]: Fixed in https://github.com/friendica/friendica/pull/3326/commits/08924b9c598fc969923fc16c3eccd90a931bbd1f

File: library/Smarty/demo/plugins/resource.mysqls.php (hunk 3 above)
Comment (Alkarex) [inline library/Smarty/demo/plugins/resource.mysqls.php]: Fixed in https://github.com/friendica/friendica/pull/3326/commits/08924b9c598fc969923fc16c3eccd90a931bbd1f

File: view/theme/frio/img/dynamic.php (hunk 4 above)
Comment (rabuzarus) [inline view/theme/frio/img/dynamic.php]: I recognized it yesterday for the first time while reviewing this PR. This file seems to be a mistake. It can be deleted.

Comment (annando) [inline view/theme/frio/img/dynamic.php]: @Alkarex could you delete that file?

Comment (Alkarex) [inline view/theme/frio/img/dynamic.php]: @annando Here you are https://github.com/friendica/friendica/pull/3326/commits/6cc4f8b0a9673c77320905f4ae18fdacb76b8f9f
